README
NAME
    CGI::IDS - PerlIDS - Perl Website Intrusion Detection System (XSS, CSRF,
    SQLI, LFI etc.)

VERSION
    Version 1.0216 - based on and tested against the filter tests of PHPIDS
    https://phpids.org rev. 1409

DESCRIPTION
    PerlIDS (CGI::IDS) is a website intrusion detection system based on
    PHPIDS https://phpids.org/ to detect possible attacks in website
    requests, e.g. Cross-Site Scripting (XSS), Cross-Site Request Forgery
    (CSRF), SQL Injections (SQLI) etc.

    It parses any hashref for possible attacks, so it does not depend on
    CGI.pm.

    The intrusion detection is based on a set of converters that convert the
    request according to common techniques that are used to hide attacks.
    These converted strings are checked for attacks by running a filter set
    of currently 68 regular expressions and a generic attack detector to
    find obfuscated attacks. For easily keeping the filter set up-to-date,
    PerlIDS is compatible to the original XML filter set of PHPIDS, which is
    frequently updated.

    Each matching regular expression has it's own impact value that
    increases the tested string's total attack impact. Using these total
    impacts, a threshold can be defined by the calling application to log
    the suspicious requests to database and send out warnings via e-mail or
    even SMS on high impacts that indicate critical attack activity. These
    impacts can be summed per IP address, session or user to identify
    attackers who are testing the website with small impact attacks over a
    time.

    You can improve the speed and the accurancy (reduce false positives) of
    the IDS by specifying an XML whitelist file. This whitelist check can
    also be processed separately by using CGI::IDS::Whitelist if you want to
    pre-check the parameters on your application servers before you send
    only the suspicious requests over to worker servers that do the complete
    CGI::IDS check.

    Download and install via CPAN:
    http://search.cpan.org/dist/CGI-IDS/lib/CGI/IDS.pm

    Report issues and contribute to PerlIDS on GitHub:
    https://github.com/hinnerk-a/perlids

SYNOPSIS
     use CGI;
     use CGI::IDS;

     $cgi = new CGI;

     # instantiate the IDS object;
     # do not scan keys, values only; don't scan PHP code injection filters (IDs 58,59,60);
     # whitelist the parameters as per given XML whitelist file;
     # All arguments are optional, 'my $ids = new CGI::IDS();' is also working correctly,
     # loading the entire shipped filter set and not scanning the keys.
     # See new() for all possible arguments.
     my $ids = new CGI::IDS(
         whitelist_file  => '/home/hinnerk/ids/param_whitelist.xml',
         disable_filters => [58,59,60],
     );

     # start detection
     my %params = $cgi->Vars;
     my $impact = $ids->detect_attacks( request => \%params );

     if ($impact > 0) {
         my_log( $ids->get_attacks() );
     }
     if ($impact > 30) {
         my_warn_user();
         my_email( $ids->get_attacks() );
     }
     if ($impact > 50) {
         my_deactivate_user();
         my_sms( $ids->get_attacks() );
     }

     # now with scanning the hash keys
     $ids->set_scan_keys(scan_keys => 1);
     $impact = $ids->detect_attacks( request => \%params );

    See examples/demo.pl in CGI::IDS module package for a running demo.

    You might want to build your own 'session impact counter' that increases
    during multiple suspicious requests by one single user, session or IP
    address.

METHODS
  new()
    Constructor. Can optionally take a hash of settings. If *filters_file*
    is not given, the shipped filter set will be loaded, *scan_keys*
    defaults to 0.

    The filter set and whitelist will stay loaded during the lifetime of the
    object. You may call `detect_attacks()' multiple times, the attack array
    (`get_attacks()') will be emptied at the start of each run of
    `detect_attacks()'.

    For example, the following is a valid constructor:

     my $ids = new CGI::IDS(
         filters_file    => '/home/hinnerk/ids/default_filter.xml',
         whitelist_file  => '/home/hinnerk/ids/param_whitelist.xml',
         scan_keys       => 0,
         disable_filters => [58,59,60],
     );

    The Constructor dies (croaks) if no filter rule could be loaded.

  detect_attacks()
     DESCRIPTION
       Parses a hashref (e.g. $query->Vars) for detection of possible attacks.
       The attack array is emptied at the start of each run.
     INPUT
       +request   hashref to be parsed
     OUTPUT
       Impact if filter matched, 0 otherwise
     SYNOPSIS
       $ids->detect_attacks(request => $query->Vars);

  set_scan_keys()
     DESCRIPTION
       Sets key scanning option
     INPUT
       +scan_keys   1 to scan keys, 0 to switch off scanning keys, defaults to 0
     OUTPUT
       none
     SYNOPSIS
       $ids->set_scan_keys(scan_keys => 1);

  get_attacks()
     DESCRIPTION
       Get an key/value/impact array of all detected attacks.
       The array is emptied at the start of each run of C<detect_attacks()>.
     INPUT
       none
     OUTPUT
       ARRAY (
         key     => '',
         value   => '',
         impact  => n,
         filters => (n, n, n, n, ...),
         tags    => ('', '', '', '', ...),
       )
     SYNOPSIS
       $ids->get_attacks();

  get_rule_description()
     DESCRIPTION
       Returns the rule description for a given rule id. This can be used for logging purposes.
     INPUT
       HASH
       + rule_id      id of rule
     OUTPUT
       SCALAR description
     EXAMPLE
       $ids->get_rule_description( rule_id => $rule_id );

XML INPUT FILES
  Filters
    This module is compatible with the PHPIDS filter set. Please find the
    latest (frequently updated) filter file from the PHPIDS Subversion
    repository at
    https://dev.itratos.de/projects/php-ids/repository/raw/trunk/lib/IDS/def
    ault_filter.xml.

    Example XML Code
     <filters>
         <filter>
             <id>1</id>
             <rule><![CDATA[(?:"+.*>)|(?:[^\w\s]\s*\/>)|(?:>")]]></rule>
             <description>finds html breaking injections including whitespace attacks</description>
             <tags>
                 <tag>xss</tag>
                 <tag>csrf</tag>
             </tags>
             <impact>4</impact>
         </filter>
         <filter>
             <id>2</id>
             <rule><![CDATA[(?:"+.*[<=]\s*"[^"]+")|(?:"\w+\s*=)|(?:>\w=\/)|(?:#.+\)["\s]*>)|(?:"\s*(?:src|style|on\w+)\s*=\s*")]]></rule>
             <description>finds attribute breaking injections including whitespace attacks</description>
             <tags>
                 <tag>xss</tag>
                 <tag>csrf</tag>
             </tags>
             <impact>4</impact>
         </filter>
     </filters>

    Used XML Tags
    * filters
        The root tag.

        * filter
            Filter item.

            * id
                Filter ID for referring in log files etc.

            * rule
                The regular expression for detection of malicious code.
                Case-insensitive; mode modifiers *i*, *m* and *s* in use.

            * description
                Description of what the filter finds.

            * tags
                Set of tags that describe the kind of attack.

                * tag
                    Currently used values are *xss*, *csrf*, *sqli*, *dt*,
                    *id*, *lfi*, *rfe*, *spam*, *dos*.

            * impact
                Value of impact, defines the weight of the attack. Each
                detection run adds the particular filter impacts to one
                total impact sum.

  Whitelist
    Using a whitelist you can improve the speed and the accurancy (reduce
    false positives) of the IDS. A whitelist defines which parameters do not
    need to undergo the expensive scanning (if their values match given
    rules and given conditions).

    Example XML Code
     <whitelist>
         <param>
             <key>scr_id</key>
             <rule><![CDATA[(?:^[0-9]+\.[0-9a-f]+$)]]></rule>
         </param>
         <param>
             <key>uid</key>
         </param>
         <param>
             <key>json_value</key>
             <encoding>json</encoding>
         </param>
         <param>
             <key>login_password</key>
             <conditions>
                 <condition>
                     <key>username</key>
                     <rule><![CDATA[(?:^[a-z]+$)]]></rule>
                </condition>
                <condition>
                    <key>send</key>
                </condition>
                <condition>
                    <key>action</key>
                    <rule><![CDATA[(?:^login$)]]></rule>
                </condition>
             </conditions>
         </param>
         <param>
             <key>sender_id</key>
             <rule><![CDATA[(?:[0-9]+\.[0-9a-f]+)]]></rule>
             <conditions>
                 <condition>
                     <key>action</key>
                     <rule><![CDATA[(?:^message$)]]></rule>
                </condition>
             </conditions>
         </param>
     </whitelist>

    Used XML Tags
    * whitelist
        The root tag.

        * param
            Parameter item. Defines the query parameter to be whitelisted.

            * key
                Parameter key.

            * rule
                Regular expression to match. If the parameter value matches
                this rule or the rule tag is not present, the IDS will not
                run its filters on it. Case-sensitive; mode modifiers *m*
                and *s* in use.

            * encoding
                Use value *json* if the parameter contains JSON encoded
                data. IDS will test the decoded data, otherwise a false
                positive would occur due to the 'suspicious' JSON encoding
                characters.

            * conditions
                Set of conditions to be fulfilled. This is the parameter
                environment in which the whitelisted parameter has to live
                in. The parameter will only be skipped if all conditions
                (and its own parameter rule) match.

                In the example XML this means: *login_password* may only be
                skipped of filtering if parameter *action* equals *login*,
                parameter *send* is present and parameter *username*
                contains only small letters.

                * condition
                    A condition to be fulfilled.

                    * key
                        Parameter key.

                    * rule
                        Regular expression to match. Missing `<rule>' means
                        *key has to be present no matter what content (can
                        even be empty)*.

    Helper methods for building and improving whitelists
     # check request
     my $impact = $ids->detect_attacks( request => $request);

     # print reasons and key/value pairs to a logfile for analysis of your application parameters.
     print LOG "filtered_keys:\n"
     foreach my $entry (@{$ids->{filtered_keys}}) {
         print LOG "\t".$entry->{reason}."\t".$entry->{key}.' => '.$entry->{value}."\n";
     }
     print LOG "non_filtered_keys:\n"
     foreach my $entry (@{$ids->{non_filtered_keys}}) {
         print LOG "\t".$entry->{reason}."\t".$entry->{key}.' => '.$entry->{value}."\n";
     }

    `$entry->{reason}' returns following reasons for skipping and
    non-skipping a value:

    `$ids->{filtered_keys}'
        * *key*: key not whitelisted
            Filtered due to missing rule set for this key.

        * *cond*: condition mismatch
            Filtered due to mismatching conditions for this key.

        * *rule*: rule mismatch
            Filtered due to mismatching rule for this key.

        * *enc*: value contains encoding
            Filtered due to containing (JSON) encoding for this key.

    `$ids->{non_filtered_keys}'
        * *empty*: empty value
            Not filtered due to empty value for this key.

        * *harml*: harmless value
            Not filtered due to harmless value string for this key.

        * *key*: key generally whitelisted
            Not filtered because the key has been generally whitelisted.

        * *r&c*: rule & conditions matched
            Not filtered due to matching rules and conditions for this key.

BUGS
    Please report any bugs or feature requests to `bug-cgi-ids at
    rt.cpan.org', or through the web interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CGI-IDS. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc CGI::IDS

    You can also look for information at:

    * RT: CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=CGI-IDS

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/CGI-IDS

    * CPAN Ratings
        http://cpanratings.perl.org/d/CGI-IDS

    * Search CPAN
        http://search.cpan.org/dist/CGI-IDS

CREDITS
    Thanks to:

    * Mario Heiderich (https://phpids.org/)
    * Christian Matthies (https://phpids.org/)
    * Ingo Bax (http://www.epublica.de/)
    * epublica GmbH (<http://www.epublica.de/>)
    * XING AG (<https://www.xing.com/>) for making this work possible and
    running PerlIDS under heavy load for a long time.
AUTHOR
    Hinnerk Altenburg, `<hinnerk at cpan.org>'

SEE ALSO

Changes
Revision history for PerlIDS

1.0216  2014-05-25
        Update docs

1.0215  2014-05-25
        Publishing repository on GitHub

1.0214  2011-08-19
        Sanitizing 'Malformed UTF-8 character (fatal)' with the new CGI::IDS::Whitelist::make_utf_8()
        Updated new URL to PHPIDS

1.0213  2011-08-05
        Catching 'Malformed UTF-8 character (fatal)' error
        Some code cleanup

1.0212  2010-06-11
        Updated filter set and converters to PHPIDS 0.6.4 (rev. 1409)

1.0211  2010-04-08
        Increased test coverage

1.021   2010-04-07
        Documentation and comment fixes

1.02    2010-04-06
        Extracted CGI::IDS::Whitelist for separately using the whitelist check
        Updated filter set and converters to PHPIDS rev. 1374

1.0118  2010-02-22
        Added XML::Parser to build_requires due to cpantesters' FAILs on some systems
        Fixed base64 warning in testForFalseAlerts

1.0117  2010-02-16
        Updated filter set and converters to PHPIDS rev. 1373
        Improvements on strip_tags()
        Thanks to Johan Nilsson for bug reports:
        - Fixed code sample in documentation
        - preg_replace(): Fixed undef warning in regex-double-execution
        - preg_replace(): Escape $ if it's not prefix of a substitution variable like $1

1.0116  2009-04-29
        Added twitter account
        Updated filter set and converters to PHPIDS rev. 1276

1.0115  2009-04-17
        Updated filter set and converters to PHPIDS rev. 1274

1.0114  2009-03-24
        Dropped utf8 pragmas, thanks to Merten Falk
        Updated filter set and converters to PHPIDS rev. 1255

1.0113  2009-03-18
        Updated filter set and converters to PHPIDS rev. 1250

1.0112  2009-03-11
        Updated filter set and converters to PHPIDS rev. 1245

1.0111  2009-03-04
        Updated filter set and converters to PHPIDS rev. 1240

1.0110  2009-02-10
        Updated filter set and converters to PHPIDS rev. 1235

1.0109  2009-02-06
        Updated filter set and converters to PHPIDS rev. 1228

1.0108  2009-01-30
        Updated filter set and converters to PHPIDS rev. 1274

1.0107  2009-01-20
        Updated filter set and converters to PHPIDS rev. 1221

1.0106  2009-01-13
        Updated filter set and converters to PHPIDS rev. 1219

1.0105  2009-01-06
        Updated filter set and converters to PHPIDS rev. 1215
        Improved Synopsis example

1.0104  2008-12-02
        Updated filter set to PHPIDS rev. 1158

1.0103  2008-11-07
        Updated to PHPIDS rev. 1152
        Fixed args handling and backslash escaping in sub preg_replace()
        Shortened substitution in sub urlencode()

1.0102  2008-11-07
        Removed XML::SAX from requirements, it didn't help with the Solaris bug.
        Improved documentation.
        Added Test::Pod::Coverage

1.0101  2008-11-06
        Fixed XMLin() bug on Solaris, now requiring XML::SAX that is independent to external OS libraries.
        Fixed links in documentation.
        Updated filter set to PHPIDS rev. 1149

1.01    2008-11-04
        Fixed README.

1.00    2008-11-04
        First CPAN release.


lib/CGI/IDS/Whitelist.pm
package CGI::IDS::Whitelist;

our $VERSION = '1.0216';

#------------------------- Notes -----------------------------------------------
# This source code is documented in both POD and ROBODoc format.
# Please find additional POD documentation at the end of this file
# (search for "__END__").
#-------------------------------------------------------------------------------

#****c* IDS::Whitelist
# NAME
#   PerlIDS Whitelist (CGI::IDS::Whitelist)
# DESCRIPTION
#   Whitelist Processor for PerlIDS (CGI::IDS)
# AUTHOR
#   Hinnerk Altenburg <hinnerk@cpan.org>
# CREATION DATE
#   2010-03-29
# COPYRIGHT
#   Copyright (C) 2010-2014 Hinnerk Altenburg
#
#   This file is part of PerlIDS.
#
#   PerlIDS is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Lesser General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   PerlIDS is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public License
#   along with PerlIDS.  If not, see <http://www.gnu.org/licenses/>.

#****

=head1 NAME

CGI::IDS::Whitelist - Whitelist Processor for PerlIDS - Perl Website Intrusion Detection System (XSS, CSRF, SQLI, LFI etc.)

=head1 DESCRIPTION

Whitelist Processor for PerlIDS (L<CGI::IDS|CGI::IDS>). Performs a basic string check and the whitelist check.
See section L<CGI::IDS/Whitelist> for details on setting up a whitelist file. CGI::IDS::Whitelist may also be
used standalone without CGI::IDS to check whether a request has suspicious parameters at all before
handing it over to CGI::IDS. This may be the case if you let worker servers do the more expensive
CGI::IDS job and only want to send over the requests that have suspicious parameters.
See L<SYNOPSIS|CGI::IDS::Whitelist/SYNOPSIS> for an example.

=head1 SYNOPSIS

 use CGI;
 use CGI::IDS::Whitelist;

 $query = new CGI;

 my $whitelist = CGI::IDS::Whitelist->new(
   whitelist_file  => '/home/hinnerk/sandbox/ids/cgi-bin/param_whitelist.xml',
 );

 my @request_keys = keys %$query->Vars;
 foreach my $key (@request_keys) {
   if ( $whitelist->is_suspicious(key => $key, request => $query->Vars ) {
     send_to_ids_worker_server( $query->Vars );
     last;
   }
 }

=head1 METHODS

=cut

#------------------------- Pragmas ---------------------------------------------
use strict;
use warnings;

#------------------------- Libs ------------------------------------------------
use XML::Simple qw(:strict);
use Carp;
use JSON::XS;
use Encode;

#------------------------- Subs ------------------------------------------------

#****m* IDS/new
# NAME
#   Constructor
# DESCRIPTION
#   Creates a Whitelist object.
#   The whitelist will stay loaded during the lifetime of the object.
#   You may call is_suspicious() multiple times, the collecting debug
#   arrays suspicious_keys() and non_suspicious_keys() will only be
#   emptied by an explizit reset() call.
# INPUT
#   HASH
#     whitelist_file  STRING  The path to the whitelist XML file
# OUTPUT
#   Whitelist object, dies (croaks) if a whitelist parsing error occurs.
# EXAMPLE
#   # instantiate object
#   my $whitelist = CGI::IDS::Whitelist->new(
#       whitelist_file  => '/home/hinnerk/sandbox/ids/cgi-bin/param_whitelist.xml',
#   );
#   # instantiate object without a whitelist, just performs a basic string check
#   my $whitelist = CGI::IDS::Whitelist->new();

#****

=head2 new()

Constructor. Can optionally take the path to a whitelist file.
If I<whitelist_file> is not given, just a basic string check will be performed.

The whitelist will stay loaded during the lifetime of the object.
You may call C<is_suspicious()> multiple times, the collecting debug
arrays C<suspicious_keys()> and C<non_suspicious_keys()> will only be
emptied by an explizit C<reset()> call.

For example, the following are valid constructors:

 my $whitelist = CGI::IDS::Whitelist->new(
     whitelist_file  => '/home/hinnerk/sandbox/ids/cgi-bin/param_whitelist.xml',
 );

 my $whitelist = CGI::IDS::Whitelist->new();

The Constructor dies (croaks) if a whitelist parsing error occurs.

=cut

sub new {
    my ($package, %args) = @_;

    # self member variables
    my $self = {
        whitelist_file      => $args{whitelist_file},
        suspicious_keys     => [],
        non_suspicious_keys => [],
    };

    # create object
    bless $self, $package;

    # read & parse XML
    $self->_load_whitelist_from_xml($self->{whitelist_file});

    return $self;
}

#****m* IDS/Whitelist/is_suspicious
# NAME
#   is_suspicious
# DESCRIPTION
#   Performs the whitelist check for a given request parameter.
# INPUT
#   HASHREF
#     + key       The key of the request parameter to be checked
#     + request   HASHREF to the complete request (for whitelist conditions check)
# OUTPUT
#   1 if you should check it with the complete filter set,
#   0 if harmless or sucessfully whitelisted.
# SYNOPSIS
#   $whitelist->is_suspicious( key => 'mykey', request => $request );
#****

=head2 is_suspicious()

 DESCRIPTION
   Performs the whitelist check for a given request parameter.
 INPUT
   HASHREF
     + key       The key of the request parameter to be checked
     + request   HASHREF to the complete request (for whitelist conditions check)
 OUTPUT
   1 if you should check it with the complete filter set,
   0 if harmless or sucessfully whitelisted.
 SYNOPSIS
   $whitelist->is_suspicious( key => 'mykey', request => $request );

=cut

sub is_suspicious {
    my ($self, %args)       = @_;
    my $key                 = $args{key};
    my $request             = $args{request};
    my $request_value       = $args{request}->{$key};
    my $contains_encoding   = 0;

    # skip if value is empty or generally whitelisted
    if ( $request_value ne '' &&
        !(  $self->{whitelist}{$key} &&
            !defined($self->{whitelist}{$key}->{rule}) &&
            !defined($self->{whitelist}{$key}->{conditions}) &&
            !defined($self->{whitelist}{$key}->{encoding})
        )
    ) {
        my $request_value_orig = $request_value;
        $request_value = $self->convert_if_marked_encoded(key => $key, value => $request_value);
        if ($request_value ne $request_value_orig) {
            $contains_encoding = 1;
        }

        $request_value = $self->make_utf_8($request_value);

        # scan only if value is not harmless
        if ( !$self->is_harmless_string($request_value) ) {
            my $attacks = {};

            if (!$self->{whitelist}{$key}) {
                # apply filters to value, not in whitelist
                push (@{$self->{suspicious_keys}}, {key => $key, value => $request_value, reason => 'key'}); # key not whitelisted
                return 1;
            }
            else {
                # check if all conditions match
                my $condition_mismatch = 0;
                foreach my $condition (@{$self->{whitelist}{$key}->{conditions}}) {
                    if (! defined($request->{$condition->{key}}) ||
                        ( defined ($condition->{rule}) && $request->{$condition->{key}} !~ $condition->{rule} )
                    ) {
                        $condition_mismatch = 1;
                    }
                }

                # Apply filters if key is not in whitelisted environment conditions
                # or if the value does not match the whitelist rule if one is set.
                # Filtering is skipped if no rule is set.
                if ( $condition_mismatch ||
                    (defined($self->{whitelist}{$key}->{rule}) &&
                    $request_value !~ $self->{whitelist}{$key}->{rule}) ||
                    $contains_encoding
                ) {
                    # apply filters to value, whitelist rules mismatched
                    my $reason = '';
                    if ($condition_mismatch) {
                        $reason = 'cond'; # condition mismatch
                    }
                    elsif (!$contains_encoding) {
                        $reason = 'rule'; # rule mismatch
                    }
                    else {
                        $reason = 'enc'; # contains encoding
                    }
                    push (@{$self->{suspicious_keys}}, {key => $key, value => $request_value, reason => $reason});
                    return 1;
                }
                else {
                    # skipped, whitelist rule matched
                    push (@{$self->{non_suspicious_keys}}, {key => $key, value => $request_value, reason => 'r&c'}); # rule & conditions matched
                }
            }
        }
        else {
            # skipped, harmless string
            push (@{$self->{non_suspicious_keys}}, {key => $key, value => $request_value, reason => 'harml'}); # harmless
        }
    }
    else {
        # skipped, empty value or key generally whitelisted
        my $reason = $request_value ? 'key' : 'empty';
        push (@{$self->{non_suspicious_keys}}, {key => $key, value => $request_value, reason => $reason});
    }
    return 0;
}

#****m* IDS/Whitelist/convert_if_marked_encoded
# NAME
#   convert_if_marked_encoded
# DESCRIPTION
#   Tries to JSON-decode and flatten a value to a plain string if the key has been marked as JSON in the whitelist.
#   Other encodings may follow in future.
# INPUT
#   HASHREF
#     + key
#     + value
# OUTPUT
#   The JSON-decoded and flattened 'value' if key is marked JSON. Plain keys and values, newline separated.
#   Untouched 'value' otherwise.
# SYNOPSIS
#   $whitelist->convert_if_marked_encoded( key => 'data', value = '{"a":"b","c":["123", 111, "456"]}');
#****

=head2 convert_if_marked_encoded()

 DESCRIPTION
   Tries to JSON-decode and flatten a value to a plain string if the key has been marked as JSON in the whitelist.
   Other encodings may follow in future.
 INPUT
   HASHREF
     + key
     + value
 OUTPUT
   The JSON-decoded and flattened 'value' if key is marked JSON. Plain keys and values, newline separated.
   Untouched 'value' otherwise.
 SYNOPSIS
   $whitelist->convert_if_marked_encoded( key => 'data', value => '{"a":"b","c":["123", 111, "456"]}');

=cut

sub convert_if_marked_encoded {
    my ($self, %args)   = @_;
    my $key             = $args{key};
    my $request_value   = $args{value};

    # If marked as JSON, try to convert from JSON to reduce false positives
    if (defined($self->{whitelist}{$key}) &&
        defined($self->{whitelist}{$key}->{encoding}) &&
        $self->{whitelist}{$key}->{encoding} eq 'json') {

        $request_value = _json_to_string($request_value);
    }
    return $request_value;
}

#****m* IDS/Whitelist/suspicious_keys
# NAME
#   suspicious_keys
# DESCRIPTION
#   Returns the set of filters that are suspicious
#   Keys are listed from the last reset() or Whitelist->new()
# INPUT
#   none
# OUTPUT
#   [ { 'value' => , 'reason' => , 'key' =>  }, { ... } ]
# SYNOPSIS
#   $whitelist->suspicious_keys();
#****

=head2 suspicious_keys()

 DESCRIPTION
   Returns the set of filters that are suspicious
   Keys are listed from the last reset() or Whitelist->new()
 INPUT
   none
 OUTPUT
   [ { 'value' => , 'reason' => , 'key' =>  }, { ... } ]
 SYNOPSIS
   $whitelist->suspicious_keys();

=cut

sub suspicious_keys {
    my ($self) = @_;
    return $self->{suspicious_keys};
}

#****m* IDS/Whitelist/non_suspicious_keys
# NAME
#   non_suspicious_keys
# DESCRIPTION
#   Returns the set of filters that have been checked but are not suspicious
#   Keys are listed from the last reset() or Whitelist->new()
# INPUT
#   none
# OUTPUT
#   [ { 'value' => , 'reason' => , 'key' =>  }, { ... } ]
# SYNOPSIS
#   $whitelist->non_suspicious_keys();
#****

=head2 non_suspicious_keys()

 DESCRIPTION
   Returns the set of filters that have been checked but are not suspicious
   Keys are listed from the last reset() or Whitelist->new()
 INPUT
   none
 OUTPUT
   [ { 'value' => , 'reason' => , 'key' =>  }, { ... } ]
 SYNOPSIS
   $whitelist->non_suspicious_keys();

=cut

sub non_suspicious_keys {
    my ($self) = @_;
    return $self->{non_suspicious_keys};
}

#****m* IDS/Whitelist/reset
# NAME
#   reset
# DESCRIPTION
#   resets the member variables suspicious_keys and non_suspicious_keys to []
# INPUT
#   none
# OUTPUT
#   none
# SYNOPSIS
#   $whitelist->reset();
#****

=head2 reset()

 DESCRIPTION
   resets the member variables suspicious_keys and non_suspicious_keys to []
 INPUT
   none
 OUTPUT
   none
 SYNOPSIS
   $whitelist->reset();

=cut

sub reset {
    my ($self) = @_;
    $self->{suspicious_keys}     = [];
    $self->{non_suspicious_keys} = [];
}

#****f* IDS/Whitelist/is_harmless_string
# NAME
#   is_harmless_string
# DESCRIPTION
#   Performs a basic regexp check for harmless characters
# INPUT
#   + string
# OUTPUT
#   BOOLEAN (pattern match return value)
# SYNOPSIS
#   $whitelist->is_harmless_string( $string );
#****

=head2 is_harmless_string()

 DESCRIPTION
   Performs a basic regexp check for harmless characters
 INPUT
   + string
 OUTPUT
   BOOLEAN (pattern match return value)
 SYNOPSIS
   $whitelist->is_harmless_string( $string );

=cut

sub is_harmless_string {
    my ($self, $string) = @_;

    $string = $self->make_utf_8($string);

    return ( $string !~ m/[^\w\s\/@!?\.]+|(?:\.\/)|(?:@@\w+)/ );
}

#****f* IDS/Whitelist/make_utf_8
# NAME
#   make_utf_8
# DESCRIPTION
#   Encodes string to UTF-8 and strips malformed UTF-8 characters
# INPUT
#   + string
# OUTPUT
#   UTF-8 string
# SYNOPSIS
#   $whitelist->make_utf_8( $string );
#****

=head2 make_utf_8()

 DESCRIPTION
   Encodes string to UTF-8 and strips malformed UTF-8 characters
 INPUT
   + string
 OUTPUT
   UTF-8 string
 SYNOPSIS
   $whitelist->make_utf_8( $string );

=cut

sub make_utf_8 {
    my ($self, $string) = @_;

    # make string UTF-8
    my $utf8_encoded = '';
    eval {
        $utf8_encoded = Encode::encode('UTF-8', $string, Encode::FB_CROAK);
    };
    if ($@) {
        # sanitize malformed UTF-8
        $utf8_encoded = '';
        my @chars = split(//, $string);
        foreach my $char (@chars) {
            my $utf_8_char = eval { Encode::encode('UTF-8', $char, Encode::FB_CROAK) }
                or next;
            $utf8_encoded .= $utf_8_char;
        }
    }
    return $utf8_encoded;
}

#****im* IDS/Whitelist/_load_whitelist_from_xml
# NAME
#   _load_whitelist_from_xml
# DESCRIPTION
#   loads the parameter whitelist XML file
#   croaks if a xml or regexp parsing error occors
# INPUT
#   whitelistfile   path + name of the XML whitelist file
# OUTPUT
#   int             number of loaded rules
# SYNOPSIS
#   $self->_load_whitelist_from_xml('/home/xyz/param_whitelist.xml');
#****

sub _load_whitelist_from_xml {
    my ($self, $whitelistfile) = @_;
    my $whitelistcnt = 0;

    if ($whitelistfile) {
        # read & parse whitelist XML
        my $whitelistxml;
        eval {
            $whitelistxml = XMLin($whitelistfile,
                forcearray  => [ qw(whitelist param conditions condition)],
                keyattr     => [],
            );
        };
        if ($@) {
            croak "Error in _load_whitelist_from_xml while parsing $whitelistfile: $@";
        }

        # convert XML structure into handy data structure
        foreach my $whitelistobj (@{$whitelistxml->{param}}) {
            my @conditionslist = ();
            foreach my $condition (@{$whitelistobj->{conditions}[0]{condition}}) {
                if (defined($condition->{rule})) {
                    # copy for error message
                    my $rule = $condition->{rule};

                    eval {
                        $condition->{rule} = qr/$condition->{rule}/ms;
                    };
                    if ($@) {
                        croak 'Error in whitelist rule of condition "' . $condition->{key} . '" for param "' . $whitelistobj->{key} . '": ' . $rule . ' Message: ' . $@;
                    }
                }
                push(@conditionslist, $condition);
            }
            my %whitelisthash = ();
            if (defined($whitelistobj->{rule})) {
                eval {
                    $whitelisthash{rule} = qr/$whitelistobj->{rule}/ms;
                };
                if ($@) {
                    croak 'Error in whitelist rule for param "' . $whitelistobj->{key} . '": ' . $whitelistobj->{rule} . ' Message: ' . $@;
                }
            }
            if (@conditionslist) {
                $whitelisthash{conditions} = \@conditionslist;
            }
            if ($whitelistobj->{encoding}) {
                $whitelisthash{encoding} = $whitelistobj->{encoding};
            }
            $self->{whitelist}{$whitelistobj->{key}} = \%whitelisthash;
            $whitelistcnt++;
        }
    }
    return $whitelistcnt;
}

#****if* IDS/Whitelist/_json_to_string
# NAME
#   _json_to_string
# DESCRIPTION
#   Tries to decode a string from JSON. Uses _datastructure_to_string().
# INPUT
#   value   the string to convert
# OUTPUT
#   value   converted string if correct JSON, the unchanged input string otherwise
# SYNOPSIS
#   IDS::Whitelist::_json_to_string($value);
#****

sub _json_to_string {
    my ($value) = @_;
    my $json_ds;
    eval {
        $json_ds = JSON::XS::decode_json($value);
    };
    if (!$@) {
        $value = _datastructure_to_string($json_ds)."\n";
    }
    return $value;
}

#****if* IDS/Whitelist/_datastructure_to_string
# NAME
#   _datastructure_to_string
# DESCRIPTION
#   Walks recursively through array or hash and concatenates keys and values to one single string (\n separated)
# INPUT
#   ref     the array/hash to convert
# OUTPUT
#   string  converted string
# SYNOPSIS
#   IDS::Whitelist::_datastructure_to_string($ref);
#****

sub _datastructure_to_string {
    my $in = shift;
    my $out = '';
    if (ref $in eq 'HASH') {
        foreach (keys %$in) {
            $out .= $_."\n";
            $out .= _datastructure_to_string($in->{$_});
        }
    }
    elsif (ref $in eq 'ARRAY') {
        foreach (@$in) {
            $out = _datastructure_to_string($_) . $out;
        }
    }
    else {
            $out .= $in."\n";
    }
    return $out;
}

1;

__END__

=head1 BUGS & SUPPORT

see L<CGI::IDS/BUGS> and L<CGI::IDS/SUPPORT>

=head1 AUTHOR

Hinnerk Altenburg, C<< <hinnerk at cpan.org> >>

=head1 SEE ALSO

L<CGI::IDS>

=head1 COPYRIGHT & LICENSE

Copyright (C) 2014 Hinnerk Altenburg (L<http://www.hinnerk-altenburg.de/>)

This file is part of PerlIDS.

PerlIDS is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

PerlIDS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with PerlIDS.  If not, see <http://www.gnu.org/licenses/>.

=cut
